### PR TITLE
Add option to set python_out flag for protoc

### PR DESCRIPTION
**Pull request by SilverTux**
- Add _split_path method to separate path from filename
  - Add _join_path method to easily concatenate paths
  - Modify proto_path_to_generated_filename, declare_out_files,
    get_out_dir to be able to handle new out_path parameter
  - Introduce new out_path attribute in py_proto_library and
    in py_grpc_library

Solution for #22033 issue.

@markdroth

### CODE REVIEW

**Comment by SilverTux** [comment]
Hi,

can I do something to proceed with this PR? Can I somehow run the 3 required checks/tests for this PR?

Cheers,
SilverTux

**Comment by gnossen** [comment]
Hi @SilverTux. Sorry this PR fell by the wayside. I think I need to understand your use case a bit better here. The Google internal version of this rule has no such flag and this has (seemingly) not caused any problems. If you want a `_pb2.py` file in a subdirectory, can you not simply put the rule in the corresponding Bazel package for that directory?

**Comment by SilverTux** [comment]
Hi @gnossen. Sorry for the late response. I can explain my problem through an example.
Imagine the following scenario:

```
|-- WS1
|   |-- BUILD.bazel (empty file)
|   |-- WORKSPACE
|   |-- cpp (cpp sources using proto files too)
|   |-- proto
|   |   |-- a.proto
|   |   \-- b.proto
|   |-- proto_BUILD.bazel
|   |-- py
|   |   |-- BUILD.bazel
|   |   \-- test.py
\-- WS2
    |-- BUILD.bazel
    |-- WORKSPACE
    \-- test_ws2.py
```

So let's assume that you would like to use (grpc)targets from an external bazel repository(WS1). With the current implementation you will get the following error:

```
INFO: Analyzed target @proto_targets//:test_grpc_grpc (51 packages loaded, 2668 targets configured).
INFO: Found 1 target...
INFO: Deleting stale sandbox base /tmp/5aa861a289df7081c651096ea3c2a3a7/sandbox
ERROR: /tmp/5aa861a289df7081c651096ea3c2a3a7/external/proto_targets/BUILD.bazel:12:1: output 'external/proto_targets/a_pb2.py' was not created
ERROR: /tmp/5aa861a289df7081c651096ea3c2a3a7/external/proto_targets/BUILD.bazel:12:1: output 'external/proto_targets/b_pb2.py' was not created
ERROR: /tmp/5aa861a289df7081c651096ea3c2a3a7/external/proto_targets/BUILD.bazel:12:1: not all outputs were created or valid
ERROR: /tmp/5aa861a289df7081c651096ea3c2a3a7/external/proto_targets/BUILD.bazel:20:1: output 'external/proto_targets/a_pb2_grpc.py' was not created
ERROR: /tmp/5aa861a289df7081c651096ea3c2a3a7/external/proto_targets/BUILD.bazel:20:1: output 'external/proto_targets/b_pb2_grpc.py' was not created
Target @proto_targets//:test_grpc_grpc failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 8.951s, Critical Path: 0.35s
INFO: 2 processes: 2 processwrapper-sandbox.

```

I've created an example repo for this problem [here](https://github.com/SilverTux/grpc_example).
If you comment out this [line](https://github.com/SilverTux/grpc_example/blob/master/WS1/proto_BUILD.bazel#L17) and [this](https://github.com/SilverTux/grpc_example/blob/master/WS1/proto_BUILD.bazel#L29) you should get an error message something similar as above.

So in short the problem exists if you would like to use bazel repositories as external dependency.
What is your opinion about this?

PS: In the example I used a patch which is slightly different than my PR, maybe I should remove this `context.attr.out_path,` line(protobuf.bzl:222) from the PR.

Cheers,
SilverTux

**Comment by SilverTux** [comment]
Hi,

did you have some time to check the above example for the problem? What do you think about it?

Cheers,
SilverTux
